# Домашние задания пятого урока.

**1. Имеются следующие данные о транзакциях в период с 01.12.2010 по 12.09.2011:**

* InvoiceNo — номер транзакции
* StockCode — код товара
* Description — описание товара
* Quantity — количество единиц товара, добавленных в заказ
* InvoiceDate — дата транзакции 
* UnitPrice — цена за единицу товара
* CustomerID — id клиента
* Country — страна, где проживает клиент

**В данной задаче ссылка для считывания данных лежит в переменной path_to_file.
Импортируйте пандас и прочитайте данные с кодировкой ISO-8859-1. Запишите полученный датафрейм в retail, а названия колонок сохраните в переменную retail_columns.**

In [3]:
import pandas as pd

path_to_file = 'data/5_data.csv.zip'
retail = pd.read_csv(path_to_file, compression='zip', encoding='ISO-8859-1')
retail_columnds = retail.columns
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

**2. Проверьте, встречаются ли в данных повторяющиеся наблюдения, и в качестве ответа укажите их количество. Если они есть, то удалите их из retail.**

In [8]:
len(retail)-len(retail.drop_duplicates())

5268

In [66]:
retail = retail.drop_duplicates()

Ответ: 5268

**3. ⭐️ Задание со звёздочкой! ⭐️**

Д**анные содержат в себе записи как об успешных транзакциях, так и об отмененных. Если пользователь отменил заказ, в начале номера транзакции (InvoiceNo) ставится C (canceled).**

**Сколько всего транзакций отменили пользователи? Каждый отдельный тип продукта считайте за отдельную транзакцию - уникальные номера считать не надо.**

In [13]:
#retail.InvoiceNo.str.startswith('C').sum()

9251

In [18]:
retail.InvoiceNo.str[0].unique()

array(['5', 'C', 'A'], dtype=object)

In [30]:
retail[retail.InvoiceNo.str[0] == 'C'].shape

(9251, 8)

Ответ: 9251

**4. Теперь отфильтруйте данные и оставьте в retail только те заказы, где Quantity > 0. В качестве ответа укажите число оставшихся строк.**

In [82]:
retail = retail.query('Quantity>0')#.shape
retail.query('Quantity>0').shape

(526054, 8)

Ответ: 526054

**5. 💀 Сложное задание! 💀**

**Посчитайте число заказов для каждого пользователя (CustomerID) из Германии (Germany). Оставьте только тех, кто совершил более N транзакций (InvoiceNo), где N – 80% процентиль. Запишите полученные id пользователей в germany_top (не весь датафрейм, только id).**

Отфильтрованные данные записаны в retail. Идентификатор заказа – InvoiceNo. Для каждого заказа в данных может встречаться более 1 строки.

In [83]:
order = retail.query('Country=="Germany"').groupby('CustomerID', as_index=False) \
                                    .InvoiceNo.nunique()
order

CustomerID  InvoiceNo
0      12426.0          1
1      12427.0          3
2      12468.0          2
3      12471.0         30
4      12472.0          7
..         ...        ...
89     13814.0          2
90     13815.0          4
91     13816.0          2
92     13817.0          2
93     14335.0          2

[94 rows x 2 columns]

In [89]:
N = order.InvoiceNo.quantile(0.8)
N

7.0

In [98]:
germ_top = order[order.InvoiceNo > N].CustomerID

**6. ⭐️ Задание со звёздочкой! ⭐️**

**Используя объект с id пользователей (germany_top), полученный на предыдущем шаге, отфильтруйте наблюдения и оставьте в данных записи только по интересующим нас юзерам. Результирующий датафрейм запишите в top_retail_germany.**

Датафрейм записан в retail, нужные id пользователей – в germany_top.

In [100]:
retail_germ_top = retail.loc[retail.CustomerID.isin(germ_top)]
retail_germ_top

InvoiceNo StockCode                          Description  Quantity  \
1109      536527     22809              SET OF 6 T-LIGHTS SANTA         6   
1110      536527     84347  ROTATING SILVER ANGELS T-LIGHT HLDR         6   
1111      536527     84945   MULTI COLOUR SILVER T-LIGHT HOLDER        12   
1112      536527     22242        5 HOOK HANGER MAGIC TOADSTOOL        12   
1113      536527     22244           3 HOOK HANGER MAGIC GARDEN        12   
...          ...       ...                                  ...       ...   
541726    581570     22139     RETROSPOT TEA SET CERAMIC 11 PC          3   
541727    581570     23077                  DOUGHNUT LIP GLOSS         20   
541728    581570     20750             RED RETROSPOT MINI CASES         2   
541729    581570     22505            MEMO BOARD COTTAGE DESIGN         4   
541730    581570      POST                              POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID  Country  
1109    12/1/2010 13:04       2.95     12662.0  Germany  
1110    12/1/2010 13:04       2.55     12662.0  Germany  
1111    12/1/2010 13:04       0.85     12662.0  Germany  
1112    12/1/2010 13:04       1.65     12662.0  Germany  
1113    12/1/2010 13:04       1.95     12662.0  Germany  
...                 ...        ...         ...      ...  
541726  12/9/2011 11:59       4.95     12662.0  Germany  
541727  12/9/2011 11:59       1.25     12662.0  Germany  
541728  12/9/2011 11:59       7.95     12662.0  Germany  
541729  12/9/2011 11:59       4.95     12662.0  Germany  
541730  12/9/2011 11:59      18.00     12662.0  Germany  

[4175 rows x 8 columns]

**7. Сгруппируйте top_retail_germany по коду товара (StockCode). Какой из продуктов добавляли в корзину чаще всего, кроме POST?** 

Note: одним заказом считается единовременная покупка любого количества товара, т.е. без учета Quantity.

In [131]:
#retail_germ_top.groupby('StockCode', as_index=False).InvoiceNo. \
#    count().sort_values('InvoiceNo', ascending=False)[0:11]

retail_germ_top.groupby('StockCode').size().sort_values(ascending=False)[0:11]

StockCode
POST     213
22326     52
22328     38
22423     34
20719     30
22961     28
22077     26
22556     26
22554     25
22728     24
20712     23
dtype: int64

Ответ: 22326

**8. Вернемся к анализу полного датасета retail. Создайте колонку Revenue с суммой покупки, используя колонки Quantity и UnitPrice.**

Отфильтрованные данные записаны в retail.

In [136]:
#retail = retail.assign(Revenue = retail.Quantity * retail.UnitPrice)
retail['Revenue'] = retail.Quantity * retail.UnitPrice
retail.head(2)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   

      InvoiceDate  UnitPrice  CustomerID         Country  Revenue  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom    15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34

**9. Для каждой транзакции (InvoiceNo), посчитайте финальную сумму заказа. В качестве ответа укажите топ-5 (InvoiceNo) по сумме заказа (через запятую с пробелом и в порядке убывания TotalRevenue)**

**Например, для следующего примера...**

InvoiceNo StockCode Quantity InvoiceDate          UnitPrice  CustomerID  Revenue

536365    85123A    6        2010-12-01 08:26:00  2.55       17850.0     15.30

536365    71053     6        2010-12-01 08:26:00  3.39       17850.0     20.34

555555    71053     1        2010-12-02 18:00:00  3.39       17850.0     3.39

**...сумма заказа будет равна:**

InvoiceNo  TotalRevenue

536365     35.64

555555     3.39

**И ответом будет:**

536365, 555555

In [147]:
retail.groupby('InvoiceNo').Revenue.sum(). \
        sort_values(ascending=False)[0:5].index.tolist()


['581483', '541431', '574941', '576365', '556444']

Ответ: 581483, 541431, 574941, 576365, 556444